pkg/cmd/roachtest/operations/backup_restore.go

```diff
@@ -69,64 +75,162 @@ outer:
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("RESTORE DATABASE %s FROM LATEST IN '%s' AS OF SYSTEM TIME '%s' WITH OPTIONS (new_db_name = '%s')", dbName, bucket, backupTS.AsOfSystemTime(), restoreDBName))
+		if err != nil {
+			o.Fatal(err)
+		}
+	} else {
+		o.Status("beginning online restore")
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("RESTORE DATABASE %s FROM LATEST IN '%s' AS OF SYSTEM TIME '%s' WITH OPTIONS (new_db_name = '%s', EXPERIMENTAL DEFERRED COPY)", dbName, bucket, backupTS.AsOfSystemTime(), restoreDBName))
+		if err != nil {
+			o.Fatal(err)
+		}
 
-	o.Status(fmt.Sprintf("verifying table district in db %s", restoreDBName))
-	backupRow, err := conn.QueryContext(ctx, fmt.Sprintf("SELECT fingerprint FROM [SHOW EXPERIMENTAL_FINGERPRINTS FROM TABLE %s.district] AS OF SYSTEM TIME '%s'", dbName, backupTS))
-	if err != nil {
-		o.Fatal(err)
+		err = waitForDownloadJob(ctx, o, c)
+		if err != nil {
+			o.Fatal(err)
+		}
 	}
-	backupRow.Next()
-	var backupFingerprint, restoreFingerprint int64
-	if err := backupRow.Scan(&backupFingerprint); err != nil {
-		o.Fatal(err)
+	o.Status(fmt.Sprintf("completed restore in %v", timeutil.Since(startTime)))
+
+	if validate {
+		o.Status(fmt.Sprintf("verifying db %s matches %s", dbName, restoreDBName))
+		sourceFingerprints, err := fingerprintutils.FingerprintDatabase(ctx, conn, dbName, fingerprintutils.AOST(backupTS), fingerprintutils.Stripped())
+		if err != nil {
+			o.Fatal(err)
+		}
+
+		// No AOST here; the timestamps are rewritten on restore. But nobody else is touching this database, so that's fine.
+		destFingerprints, err := fingerprintutils.FingerprintDatabase(ctx, conn, restoreDBName, fingerprintutils.Stripped())
+		if err != nil {
+			o.Fatal(err)
+		}
+
+		if !reflect.DeepEqual(sourceFingerprints, destFingerprints) {
+			o.Fatalf("backup and restore fingerprints do not match: %v != %v", sourceFingerprints, destFingerprints)
+		}
 	}
 
-	restoredRow, err := conn.QueryContext(ctx, fmt.Sprintf("SELECT fingerprint FROM [SHOW EXPERIMENTAL_FINGERPRINTS FROM TABLE %s.district]", restoreDBName))
+	return &backupRestoreCleanup{db: restoreDBName}
+}
+
+func waitForDownloadJob(ctx context.Context, o operation.Operation, c cluster.Cluster) error {
+	o.Status(`Begin tracking online restore download phase completion`)
+	// Wait for the job to succeed.
+	pollingInterval := 10 * time.Second
+	succeededJobTick := time.NewTicker(pollingInterval)
+	defer succeededJobTick.Stop()
+	done := ctx.Done()
+	conn, err := c.ConnE(ctx, o.L(), c.Node(1)[0])
 	if err != nil {
-		o.Fatal(err)
+		return err
 	}
-	restoredRow.Next()
-	if err := restoredRow.Scan(&restoreFingerprint); err != nil {
-		o.Fatal(err)
+	defer conn.Close()
+	for {
+		select {
+		case <-done:
+			return ctx.Err()
+		case <-succeededJobTick.C:
+			var status string
+			if err := conn.QueryRow(`SELECT status FROM [SHOW JOBS] WHERE job_type = 'RESTORE' ORDER BY created DESC LIMIT 1`).Scan(&status); err != nil {
```

Comment:
are you sure that no one will run a restore after we begin running the restore this helper is tracking? To be totally sure, you could grab the download job id in the return results of the link phase restore cmd, and then poll for the status of the download job id using this roachtest [utility](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/tests/jobs_util.go#L225). See this patch for grabbing the download job id https://github.com/cockroachdb/cockroach/pull/121388

Reply:
done, thanks